Implement `time`

envp opened the pull request:
This PR will track building `time`

Partially fixes #53

envp commented:
I tried running `cargo fmt -p time -- --check` locally, but I'm not able to reproduce the CI's results. What config are you using?

GrayJack commented:
Sorry, but could you change the commit messages to use the [project standard commit message style](https://github.com/GrayJack/coreutils/blob/dev/CONTRIBUTING.md#git-commit-messages), I decided on this standard because the project is a workspace crate of several other crates, so it's hard to know what commit is related to what crate.

Formatting is using nightly because several features in the rustfmt are still as unstable feature.

GrayJack commented on time/Cargo.toml:
```diff
@@ -0,0 +1,15 @@
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[dependencies]
+clap = { version = "^2.33.0", features = ["yaml", "wrap_help"] }
```
Looks like you're using yaml feature of clap. We dropped that due some problems with it. You can find the template used now [here](https://github.com/GrayJack/coreutils/tree/dev/.template)

GrayJack commented on Cargo.toml:
```diff
@@ -22,4 +22,5 @@ members = [
     "tty",
     "wc",
     "yes",
+    "time",
```
Can you add this to all toml files related to the OS that this utility can be compiled? The reason behind this is because no all OS works the same, and sometimes (like fuchsia and haiku) doesn't even has a way to do something. The CI also uses the toml files os respective OS to build and test the crates

GrayJack commented on time/src/rusage.rs:
```diff
@@ -0,0 +1,118 @@
+}
+
+fn duration_from_timeval(tv: timeval) -> Duration {
+    let seconds = u64::try_from(tv.tv_sec).expect("Unable to convert tv_sec to u64");
+    let micros = u64::try_from(tv.tv_usec).expect("Unable to convert tv_usec to u64");
```
There is any case where this conversion will fail? Because panic exit with a different message than the tools messages.
Or is it the case where failing to convert in any OS with this tool should be considered a bug in the utility?

envp replied:
I don't expect the conversion to fail, since `libc::getrusage` won't return negative values for the timeval struct.

The latter is true. I would think it's a bug in `time`, since this is an implementation detail.

> Because panic exit with a different message than the tools messages

I'm not sure I follow, this message corresponds to an internal error, if that helps clarify my usage.

GrayJack replied:
>  don't expect the conversion to fail, since `libc::getrusage` won't return negative values for the timeval struct.
> 
> The latter is true. I would think it's a bug in `time`, since this is an implementation detail.

In this case panic is fine



> I'm not sure I follow, this message corresponds to an internal error, if that helps clarify my usage.

I was just worried that panic exit may happen in a non-bug case where the error message should be similar to user error messages like the other tools

GrayJack reviewed (commented):
I left some comments for us to talk.

I feel that `rusage.rs` could be a module in the coreutils_core crate, but that can be done at any time, even in another PR.

envp committed:
General: Add `time` binary to toml files of OSes

envp committed:
Time: Scaffold code from template

- Initial structure taken from .template with tool specific info filled
  in

envp committed:
Time: cli: Add -p flag, and accept a command to run

envp committed:
Time: cli: Rename struct s/Args/TimeOpts

envp committed:
Core: Create, export module resource

- Wraps getrusage(2)

envp committed:
Time: cli: Combine exec_args and executable into a single field

envp committed:
Time: subprocess: Create subprocess module

envp committed:
Time: Parse CLI opts, run subprocess and extract resource usage

envp committed:
Time: cli: Fix bug in argument specification

No longer use number_of_values(true) as that implies an option

Also use "ARGUMENT" to match from clap

envp committed:
Time: cli: Refactor and add test cases

- Refactor enum declaration to use `Self`
- Add test cases for `TimeOpts` creation
- Rename 'OutputFormat' to 'OutputFormatter'

envp committed:
Time: Break up CLI into flags, output modules

envp committed:
Time: Implement `POSIX.2` output formatting and macos (FreeBSD) default

envp committed:
Time: Format code

envp committed:
Time: output: Move platform specific details into their own modules

envp committed:
Time: subprocess: Follow `POSIX.2` spec for error codes

Unspecified errors are arbitrarily mapped to the 1-125 range

envp committed:
Time: cli: Drop non-standard `--posix`

envp committed:
Time: drop macos specific code

envp committed:
Time: Custom get_rusage() for FuchsiaOS

Returns a zeroed RUsage struct

envp commented:
@GrayJack I've removed this from Fuchsia for now, as there `getrusage` isn't available.

envp commented:
@GrayJack 

A few questions:
1. Fuchsia doesn't support the `getrusage` syscall. Do you think it's best to remove the target altogether?
2. What do you think about getting `time` to be POSIX compliant first, and then adding OS-specific extensions? That would make both review and development easier as we can address the exceptions on a case by case basis.
3. I'm not able to replicate the formatting requirements locally (or inside an ubuntu docker) with nightly toolchain. Both of those seem to pass for me. What's the best way to go about matching the formatting standards?

GrayJack commented:
>     1. Fuchsia doesn't support the `getrusage` syscall. Do you think it's best to remove the target altogether?

Yes, if is something required to support POSIX standard. We have this many toml files for 2 reasons

1. Sometimes the API used on different systems are different and we want to at least build for some systems before moving to support other systems
2. Many tools can be implemented for non UNIX systems, but has some UNIXisms, like Fuchsia, Redox, Haiku, etc, but since not all tools can be implemented on them, having separated toml files for them allow us to check only the ones that are possible to build.


>     2. What do you think about getting `time` to be POSIX compliant first, and then adding OS-specific extensions? That would make both review and development easier as we can address the exceptions on a case by case basis.

That is totally fine, in most tools we only require for POSIX compliance to consider the tool complete, and extensions, being OS specific or a available on all OSes, are greatly accepted additions


>     3. I'm not able to replicate the formatting requirements locally (or inside an ubuntu docker) with nightly toolchain. Both of those seem to pass for me. What's the best way to go about matching the formatting standards?

That's odd, It is working fine for me just doing `cargo fmt`. In the worst case scenario, if you gave me permission to push changes to your PR, I can format before merging.

envp committed:
Time: Format source code

envp commented:
> > ```
> > 1. Fuchsia doesn't support the `getrusage` syscall. Do you think it's best to remove the target altogether?
> > ```
> 
> Yes, if is something required to support POSIX standard. We have this many toml files for 2 reasons
> 
>     1. Sometimes the API used on different systems are different and we want to at least build for some systems before moving to support other systems
> 
>     2. Many tools can be implemented for non UNIX systems, but has some UNIXisms, like Fuchsia, Redox, Haiku, etc, but since not all tools can be implemented on them, having separated toml files for them allow us to check only the ones that are possible to build.
> 

Gotcha, it's been commented out of `Fuchsia.toml` for now. Is that all that is required?

> > ```
> > 2. What do you think about getting `time` to be POSIX compliant first, and then adding OS-specific extensions? That would make both review and development easier as we can address the exceptions on a case by case basis.
> > ```
> 
> That is totally fine, in most tools we only require for POSIX compliance to consider the tool complete, and extensions, being OS specific or a available on all OSes, are greatly accepted additions

Great, thanks! I'll make an issue with the TODO items from this PR, and set up machinery for platform extensions in a follow-up PR.
 
> > ```
> > 3. I'm not able to replicate the formatting requirements locally (or inside an ubuntu docker) with nightly toolchain. Both of those seem to pass for me. What's the best way to go about matching the formatting standards?
> > ```
> 
> That's odd, It is working fine for me just doing `cargo fmt`. In the worst case scenario, if you gave me permission to push changes to your PR, I can format before merging.

Are you using an ubuntu/debian machine?

Also, I think the build is failing because it seems there is a crate called `time` on crates.io already (there's a complaint of potential segfault)

GrayJack commented:
> Is that all that is required?

Yes 😄 

> Great, thanks! I'll make an issue with the TODO items from this PR, and set up machinery for platform extensions in a follow-up PR.

That's great to hear, we're always happy to receive contributions 😄 



> Are you using an ubuntu/debian machine?

I use Manjaro, it is Arch-based

> Also, I think the build is failing because it seems there is a crate called `time` on crates.io already (there's a complaint of potential segfault)

Yes, it's the crate we use to handle time in some other tools. It's a possible segfault by using a non-thread-safe function call on UNIX platforms, but since for now all tools are single threaded, it's fine (for now).
But cargo-deny check don't block merges.

GrayJack commented on coreutils_core/src/os/resource.rs:
```diff
@@ -0,0 +1,109 @@
+    pub num_signals:   u64,
+}
+
+impl RUsage {
+    fn from(ru: rusage) -> Self {
+        RUsage {
+            timing: Timing { user_time: ru.ru_utime, sys_time: ru.ru_stime },
+            mem:    MemoryUsage {
+                max_rss: ru.ru_maxrss as u64,
+                num_minor_page_flt: ru.ru_minflt as u64,
+                num_major_page_flt: ru.ru_majflt as u64,
+                num_vol_ctx_switch: ru.ru_nvcsw as u64,
+                num_invol_ctx_switch: ru.ru_nivcsw as u64,
+                shared_mem_size: ru.ru_ixrss as u64,
+                unshared_data_size: ru.ru_idrss as u64,
+                unshared_stack_size: ru.ru_isrss as u64,
+                num_swaps: ru.ru_nswap as u64,
+            },
+            io:     IOUsage {
+                num_block_in:  ru.ru_inblock as u64,
+                num_block_out: ru.ru_oublock as u64,
+                num_sock_recv: ru.ru_msgrcv as u64,
+                num_sock_send: ru.ru_msgsnd as u64,
+                num_signals:   ru.ru_nsignals as u64,
+            },
+        }
+    }
+}
```
This would be better as a `impl From<rusage> for RUsage`

GrayJack reviewed (commented):
Besides the from function that should be a `From` impl, the only other thing that caught my attention are the public fields of the structs in `coreutils_core::os::resource` considering that coreutils_core is suppose to be of a medium to high level of API abstraction for the raw API used to implement the utilities, so the access to the field is suppose to be done using functions that return references to the fields, but I think that is kinda of a big change that can be addressed in another PR.

Besides that, nothing else seems out of place.

envp commented:
> Besides the from function that should be a `From` impl, the only other thing that caught my attention are the public fields of the structs in `coreutils_core::os::resource` considering that coreutils_core is suppose to be of a medium to high level of API abstraction for the raw API used to implement the utilities, so the access to the field is suppose to be done using functions that return references to the fields, but I think that is kinda of a big change that can be addressed in another PR.
> 
> Besides that, nothing else seems out of place.

I see, I was considering implementing functions for "derived" properties (per GNU time's `%` modifiers, like [`"%D"`](http://git.savannah.gnu.org/cgit/time.git/tree/src/time.c#n397)), wouldn't it be prudent to keep the getter/setters to a minimum and use struct fields (only using functions when they summarize some kind of computation)? (especially since the `rusage` interface is unlikely to change).

Ack on implementing the `From<...>` trait didn't realize that was a thing :)

envp committed:
Core: resource: Implement Conversion trait for rusage -> RUsage

GrayJack commented on coreutils_core/src/os/resource.rs:
```diff
@@ -6,6 +6,8 @@ use super::TimeVal;
 use libc::getrusage;
 use libc::{c_int, rusage, RUSAGE_CHILDREN, RUSAGE_SELF};
 
+use std::convert::From;
```
From is in the Rust prelude, so you don't need to pull it into the scope

GrayJack commented:
> wouldn't it be prudent to keep the getter/setters to a minimum and use struct fields (only using functions when they summarize some kind of computation)? (especially since the `rusage` interface is unlikely to change).

That depends, I personally think about the contract I want with the library usage. For example, in the `Utmpx`, I made with methods to get a immutable reference to all the possible fields (unless if return type is `Copy`), because it's a struct made to the user only get the system information, but not modify it, so I made an API that even if the user creates a mutable instance of `Utmpx`, it cannot modify whatever it holds.

But since in this case all possible fields are `Copy`, maybe it is a ok thing. But still, a user that creates a mutable instance of `RUsage` can by mistake modify it (maybe it's just me that are paranoid with it because I lost count the amount of times I did exactly that before Rust and when I was still a inexperienced rustacean). I'll say let's roll like it is now and moving forward well see if it will be a need to make everything private and create methods to access the values.

envp committed:
Time: subprocess: Drop benchmark like test

This test is truly arbitrary and more of a benchmark than a test

envp committed:
Core: resource: Drop import as std::convert::From is in the Prelude

GrayJack commented:
bors r+
